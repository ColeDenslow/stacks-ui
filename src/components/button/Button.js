import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';

import LoadingSpinner from '../loadingspinner/LoadingSpinner';

import './Button.css';

function Button(props) {
  const { block, children, className, disabled, href, loading, 'aria-label': ariaLabel, primary, secondary, size, target, title, type,
    ...otherProps } = props;

  const classNames = cn([
    'stx-button',
    { 'stx-button--primary': primary },
    { 'stx-button--secondary': secondary && !primary },
    { 'stx-button--block': block },
    { 'stx-button--small': size === 'small' },
    { 'stx-button--large': size === 'large' },
    { 'stx-button--is-disabled': disabled },
    { 'stx-button--is-loading': loading },
    className,
  ]);

  const ButtonTag = href && !disabled ? 'a' : 'button';
  const buttonType = (ButtonTag === 'button' && type) ? { type } : {};

  return (
    <ButtonTag
      title={title}
      className={classNames}
      href={href}
      target={target}
      disabled={disabled}
      aria-disabled={disabled}
      aria-label={ariaLabel}
      aria-busy={loading}
      {...buttonType}
      {...otherProps}
    >
      <span className='stx-button__text'>{children}</span>
      <LoadingSpinner className='stx-button__spinner' />
    </ButtonTag>
  );
}

Button.defaultProps = {
  title: null,
  className: '',
  href: null,
  target: null,
  size: null,
  type: null,

  'aria-label': null,
  primary: false,
  disabled: false,
  loading: false,
  block: false,
};

Button.propTypes = {
  /** Value given to the title attribute */
  title: PropTypes.string,
  /** Value appended to the class attribute */
  className: PropTypes.string,
  /**
   * Value given to the href attribute
   * Note: The Button will render as an <a> tag if this is provided
   */
  href: PropTypes.string,
  /** Value given to the target attribute */
  target: PropTypes.string,
  /** Sets the size of the generated button */
  size: PropTypes.oneOf(['', 'small', 'large']),
  /** Sets the type of the generated button */
  type: PropTypes.oneOf(['submit', 'reset', 'button']),

  /** String indicating the aria label */
  'aria-label': PropTypes.string,
  /** Boolean indicating whether the button should render as primary */
  primary: PropTypes.bool,
  /** Boolean indicating whether the button should render as secondary */
  secondary: PropTypes.bool,
  /** Boolean indicating whether the button should render as disabled */
  disabled: PropTypes.bool,
  /** Boolean indicating whether the button should render a loading spinner */
  loading: PropTypes.bool,
  /** Boolean indicating whether the button should render a block button */
  block: PropTypes.bool,

  /** Child elements */
  children: PropTypes.node,
};

export default Button;
