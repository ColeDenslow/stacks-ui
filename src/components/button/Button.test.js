import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from './Button';

describe('Button', () => {
  const onClick = () => 'onClick';
  const onFocus = () => 'onFocus';
  const onBlur = () => 'onBlur';
  const onMouseEnter = () => 'onMouseEnter';
  const onMouseLeave = () => 'onMouseLeave';
  const onTouchEnd = () => 'onTouchEnd';

  test('Button renders with basic props', () => {
    const output = mount(
      <Button
        className='something'
        onClick={onClick}
        onFocus={onFocus}
        onBlur={onBlur}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onTouchEnd={onTouchEnd}
      >
        {'Button text'}
      </Button>
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Button renders secondary', () => {
    const output = mount(
      <Button secondary>
        {'Button text'}
      </Button>
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Button renders primary', () => {
    const output = mount(
      <Button primary>
        {'Button text'}
      </Button>
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Button renders disabled', () => {
    const output = mount(
      <Button disabled>
        {'Button text'}
      </Button>
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Button renders loading', () => {
    const output = mount(
      <Button loading>
        {'Button text'}
      </Button>
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Button renders with additional accessibility/custom props', () => {
    const output = mount(
      <Button
        className='something'
        onClick={onClick}
        onFocus={onFocus}
        onBlur={onBlur}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onTouchEnd={onTouchEnd}
        role='button'
        aria-invalid={false}
        data-additional-prop='additional-prop-test'
      >
        {'Button text'}
      </Button>
    );

    expect(toJson(output)).toMatchSnapshot();
  });
});
