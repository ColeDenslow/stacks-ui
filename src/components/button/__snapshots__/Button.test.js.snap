// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button renders disabled 1`] = `
<Button
  block={false}
  className=""
  disabled={true}
  href={null}
  loading={false}
  primary={false}
  size={null}
  target={null}
  title={null}
  type={null}
>
  <button
    aria-busy={false}
    aria-disabled={true}
    className="stx-button stx-button--is-disabled"
    disabled={true}
    href={null}
    target={null}
    title={null}
  >
    <span
      className="stx-button__text"
    >
      Button text
    </span>
    <LoadingSpinner
      className="stx-button__spinner"
    >
      <span
        className="stx-loading-spinner stx-button__spinner"
      />
    </LoadingSpinner>
  </button>
</Button>
`;

exports[`Button Button renders loading 1`] = `
<Button
  block={false}
  className=""
  disabled={false}
  href={null}
  loading={true}
  primary={false}
  size={null}
  target={null}
  title={null}
  type={null}
>
  <button
    aria-busy={true}
    aria-disabled={false}
    className="stx-button stx-button--is-loading"
    disabled={false}
    href={null}
    target={null}
    title={null}
  >
    <span
      className="stx-button__text"
    >
      Button text
    </span>
    <LoadingSpinner
      className="stx-button__spinner"
    >
      <span
        className="stx-loading-spinner stx-button__spinner"
      />
    </LoadingSpinner>
  </button>
</Button>
`;

exports[`Button Button renders primary 1`] = `
<Button
  block={false}
  className=""
  disabled={false}
  href={null}
  loading={false}
  primary={true}
  size={null}
  target={null}
  title={null}
  type={null}
>
  <button
    aria-busy={false}
    aria-disabled={false}
    className="stx-button stx-button--primary"
    disabled={false}
    href={null}
    target={null}
    title={null}
  >
    <span
      className="stx-button__text"
    >
      Button text
    </span>
    <LoadingSpinner
      className="stx-button__spinner"
    >
      <span
        className="stx-loading-spinner stx-button__spinner"
      />
    </LoadingSpinner>
  </button>
</Button>
`;

exports[`Button Button renders secondary 1`] = `
<Button
  block={false}
  className=""
  disabled={false}
  href={null}
  loading={false}
  primary={false}
  secondary={true}
  size={null}
  target={null}
  title={null}
  type={null}
>
  <button
    aria-busy={false}
    aria-disabled={false}
    className="stx-button stx-button--secondary"
    disabled={false}
    href={null}
    target={null}
    title={null}
  >
    <span
      className="stx-button__text"
    >
      Button text
    </span>
    <LoadingSpinner
      className="stx-button__spinner"
    >
      <span
        className="stx-loading-spinner stx-button__spinner"
      />
    </LoadingSpinner>
  </button>
</Button>
`;

exports[`Button Button renders with additional accessibility/custom props 1`] = `
<Button
  aria-invalid={false}
  block={false}
  className="something"
  data-additional-prop="additional-prop-test"
  disabled={false}
  href={null}
  loading={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  primary={false}
  role="button"
  size={null}
  target={null}
  title={null}
  type={null}
>
  <button
    aria-busy={false}
    aria-disabled={false}
    aria-invalid={false}
    className="stx-button something"
    data-additional-prop="additional-prop-test"
    disabled={false}
    href={null}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    role="button"
    target={null}
    title={null}
  >
    <span
      className="stx-button__text"
    >
      Button text
    </span>
    <LoadingSpinner
      className="stx-button__spinner"
    >
      <span
        className="stx-loading-spinner stx-button__spinner"
      />
    </LoadingSpinner>
  </button>
</Button>
`;

exports[`Button Button renders with basic props 1`] = `
<Button
  block={false}
  className="something"
  disabled={false}
  href={null}
  loading={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  primary={false}
  size={null}
  target={null}
  title={null}
  type={null}
>
  <button
    aria-busy={false}
    aria-disabled={false}
    className="stx-button something"
    disabled={false}
    href={null}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    target={null}
    title={null}
  >
    <span
      className="stx-button__text"
    >
      Button text
    </span>
    <LoadingSpinner
      className="stx-button__spinner"
    >
      <span
        className="stx-loading-spinner stx-button__spinner"
      />
    </LoadingSpinner>
  </button>
</Button>
`;
